### ISSUE TITLE
multiple edges won't be merged

### ISSUE THREAD

**Issue by ChrisHH** [gephi/gephi#1251]
I am using gephi 0.9.0 (latest version) on my Macbook Pro.
It worked fine so far.

After doing the latest updates I observed the problem, that multiple edges won't be merged (and up-weighted) any longer. It looks like, that multiple edges will be ignored after their first occurrence and the weight is set to 1.

Best
Chris

**Comment by eduramiba** [comment]
What kind of import are you doing? Spreadsheet import? It now in fact does not sum weights when duplicated edges are found. Instead it not overwrites the edge attributes with the new values.

This has been changed because, in the near future, spreadsheet import will support multi-graphs.
If the old weight summing was useful for you, I can add an option to enable it back.

**Comment by ChrisHH** [comment]
I am doing the CSV import via data laboratory. 
For now the old weight summing is very useful to me, as the multiple edges are thicker in the graph. 
Yes, maybe an option can be provided within the import dialogue!?

**Comment by eduramiba** [comment]
Sure, I will add an option to sum weights like before when duplicated edges are found.

**Comment by ChrisHH** [comment]
Great, thank you!

**Comment by eduramiba** [comment]
Also need to update the wizard text, now it is incorrect. It still says that attributes will be ignored and weights summed.

**Comment by ChrisHH** [comment]
Yes, that was actually confusing me and lead me to re-install gephi.

**Comment by GitHubFDC** [comment]
It's my understanding that one of the benefits of Gephi 0.9 is that it handles multi-edges, which is a great benefit to have for certain types of modeling.  It sounds like the fix will be to add an option to either a) allow multi-edges, or b) provide an option to allow the import function to merge duplicate edges and sum their weights.  This sounds like a great solution, provided Gephi doesn't lose the ability to handle more than one edge that connects two nodes.  -Thanks!

**Comment by eduramiba** [comment]
There is an issue to support multi-edges in spreadsheet import (#1248). But that will need, like gexf import, a 'Kind' column with different values for different edges.

Repeated edges with the same kind will still be merged, and their weights should be summed. So that's why we actually have a bug here and I need to get the weight summing feature back.

**Comment by duncdrum** [comment]
@eduramiba You can see where I am with ````@kind```` [here](https://github.com/duncdrum/gexf-1)

**Comment by GitHubFDC** [comment]
One other quick question with multi-edges in Gephi 0.9--more specifically, dynamic multi-edges.

Suppose one imports a list of dynamic edges--including such entries as:
    Id     Source    Target    Weight    Type           Duration            Kind
    E1       N1           N2          1.0       directed      <1.0, 2.0>         "works_for"
    E2       N1           N2          2.0       directed      <3.5, 5.0>         "works_for"

Would E1 and E2 be merged as a SINGLE edge that is "on" during (t=1.0 and t=2.0 with weight 1.0) and again during (t=3.5 and t=5.0 with weight 2.0), or would Gephi 0.9 import these as two separate edges.  Or, would these be seen as two separated edges, despite the "kind" being identical between them. Or, can the import function handle the merging of dynamic multi-edges?

Thanks!

**Comment by duncdrum** [comment]
@GitHubFDC ````@kind```` works more like ````@type```` in this respect. An edge can appear at multiple times between identical nodes independent of its kind. So yes these would be two separate edges, since they have separate ````@id````. They just happen to be of the same kind. To makes this one edge you would have to do something like this. 
````xml
[...]
<edge id="1" source="N1" target="N2" weight="1.0, 2.0, 3.5, 5.0" type="directed" kind="works_for"/>
````
and use listDouble as the datatype for the weight attribute. You can follow the work on this here:
https://github.com/gephi/gexf .

If you want to suggest a merge on kind feature for Gephi's data import you can open a new issue.

**Comment by eduramiba** [comment]
Actually there can be no two edges with the same kind, source and target even if they have a different id. Core graph API does not allow this.

About spreadsheet import, we will only sum edge weights when duplicated (same source, target and kind), I don't think we will merge time intervals, in order to keep things simple and unambiguous. You should write the whole interval interval in one place if you want that (<[1.0, 2.0]; [3.5, 5.0]>).

The related issue for this is https://github.com/gephi/gephi/issues/1248

**Comment by duncdrum** [comment]
Thanks @eduramiba for clarifying, i ll have to make sure that the schema checks for this.

**Comment by eduramiba** [comment]
I don't think the gexf schema needs to check for this. That will be done by the gexf importer, right? @mbastian knows this better

**Comment by duncdrum** [comment]
Yes, but if people construct gexf files outside of Gephi there should be a validation errror, explaining that ````@id```` is not enough to determine an edge in this case.

**Comment by jpoko** [comment]
This still does not work (summing of edge weights). I am having this problem in Gephi 0.8 and 0.9 (Windows 7), whether I import/open and append .gexf files or .csv files. The edge weights are not being added together. The Source and Target are the same, though the edge ID differs (this should not matter though, correct? It should be identifying matching edges based on the Source and Target information). Any help is much appreciated!
Jen

**Comment by eduramiba** [comment]
Hi, this should work with latest updates.
Can you post a sample of your file?

Yes, the different id should not matter if source and target are the same.

**Comment by jpoko** [comment]
Here are the 2 .gexf files. Both are directed networks that have an edge between 6 (source) and 9 (target). For file 004, the edge ID is 1 and for file 007 the edge ID is 8. When appended, the edge weight remains 1 and the edge ID remains 1 or 8 (depending on which file was opened first). 

Furthermore, not all the edges are being created in the merged/appended graph and an error appears: 

java.lang.IllegalArgumentException: The edge id already exist
	at org.gephi.graph.impl.EdgeStore.checkIdDoesntExist(EdgeStore.java:901)
	at org.gephi.graph.impl.EdgeStore.add(EdgeStore.java:468)
	at org.gephi.graph.impl.GraphStore.addEdge(GraphStore.java:147)
	at org.gephi.io.processor.plugin.DefaultProcessor.process(DefaultProcessor.java:202)
	at org.gephi.io.processor.plugin.AppendProcessor.process(AppendProcessor.java:88)
	at org.gephi.io.importer.impl.ImportControllerImpl.process(ImportControllerImpl.java:292)
	at org.gephi.desktop.importer.DesktopImportControllerUI.finishImport(DesktopImportControllerUI.java:630)
	at org.gephi.desktop.importer.DesktopImportControllerUI.access$200(DesktopImportControllerUI.java:101)
	at org.gephi.desktop.importer.DesktopImportControllerUI$6.run(DesktopImportControllerUI.java:282)
[catch] at org.gephi.utils.longtask.api.LongTaskExecutor$RunningLongTask.run(LongTaskExecutor.java:275)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)



[network_004_007_early_theme.zip](https://github.com/gephi/gephi/files/118438/network_004_007_early_theme.zip)

**Comment by mbastian** [comment]
> The Source and Target are the same, though the edge ID differs (this should not matter though, correct? It should be identifying matching edges based on the Source and Target information). 

No, the append algorithm only uses the edge ID at the moment and doesn't "merge" parallel edges. The edge weight strategy for parallel edges is only applied to the file you're importing, not when you're appending. In other words if your .gexf file contains parallel edges (with different ids) they will be merged based on the strategy you choose. The append algorithm will add missing edges/nodes to the graph and replace attributes if new values are provided but doesn't merge. That said it shouldn't try to add the edge if it already exists with the same id so there might be an issue there. I'll take a look.

**Comment by eduramiba** [comment]
Right, also have in mind this issue was about spreadsheet import and not gexf.

**Comment by jpoko** [comment]
Is it possible to add the ability to have parallel edges be merged and weight added when files are appended? Essentially ignoring edge ID and only using the source, target, and edge weight information for a file imported on its own and when additional files are appended.

**Comment by mbastian** [comment]
@jpoko Yes, that should be possible but I discovered a bug, tracked here #1347, I'll fix it and it should work.

**Comment by jpoko** [comment]
Thank you - much appreciated!

**Comment by otichy** [comment]
Thanks for the fix, but the problem is back with 0.9.1 and 0.9.1-SNAPSHOT. On CSV import the duplicate edges are not merged and their weight remains 1.

**Comment by 3ruce** [comment]
I've also seen this today - I had to import a gdf file into 0.8.2 where edge weights were calculated, export both tables from the data lab and then reimport to 0.9.1

**Comment by eduramiba** [comment]
Can you share the CSV file with the problem?

**Comment by 3ruce** [comment]
@eduramiba Hi Eduardo, do you mean my gdf file or @otichy's CSV?

**Comment by eduramiba** [comment]
Oh this seems a different problem now. Since with 0.9.1 real parallel edges (same source, target and kind) are allowed, something must have changed in the graphstore API so the spreadsheet importer is not detecting that an edge already exists and does not sum weights, but actually creates another edge with same source and target (which should be the default behaviour in the future, but still have an option to sum weights).

I could reproduce it with this file: [test.zip](https://github.com/gephi/gephi/files/151142/test.zip)

Will have to investigate this, thanks for the report.

**Comment by otichy** [comment]
Just in case more examples are needed, here is my CSV. Thanks!

[pceec-concordance.zip](https://github.com/gephi/gephi/files/151904/pceec-concordance.zip)

**Comment by zebraengine** [comment]
I noticed that in the 0.9.2 development nightly - isn't merging or providing a check box to merge duplicate edges that have the same ID pairs. Am i suppose to enable this prior to utilizing the data laboratory import process?

**Comment by jpoko** [comment]
I installed the newest Gephi version 0.9.1 and tried to merge/append the graphs that I was working with previously (comment above on Feb 4) and it is having the same problem - the edge weight does not increase and it returns the error below. I have it selected to append the graph and the edge merge strategy set to 'sum'. 


java.lang.IllegalArgumentException: The edge id already exist
	at org.gephi.graph.impl.EdgeStore.checkIdDoesntExist(EdgeStore.java:1011)
	at org.gephi.graph.impl.EdgeStore.add(EdgeStore.java:541)
	at org.gephi.graph.impl.GraphStore.addEdge(GraphStore.java:163)
	at org.gephi.io.processor.plugin.DefaultProcessor.process(DefaultProcessor.java:210)
	at org.gephi.io.processor.plugin.AppendProcessor.process(AppendProcessor.java:88)
	at org.gephi.io.importer.impl.ImportControllerImpl.process(ImportControllerImpl.java:292)
	at org.gephi.desktop.importer.DesktopImportControllerUI.finishImport(DesktopImportControllerUI.java:638)
	at org.gephi.desktop.importer.DesktopImportControllerUI.access$200(DesktopImportControllerUI.java:102)
	at org.gephi.desktop.importer.DesktopImportControllerUI$6.run(DesktopImportControllerUI.java:286)
[catch] at org.gephi.utils.longtask.api.LongTaskExecutor$RunningLongTask.run(LongTaskExecutor.java:274)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)

**Comment by judithmp** [comment]
Did I get it right that this issue is touching different subitems?
a) spreadsheet import
b) gexf import
c) multiple networks
...

For a) I can confirm @jpoko (on 12/4):
I tried to import a csv with duplicated edges - neither in 0.9.1 nor in 0.9.2 the edges were merged and their weight summed. (Mac OS 10.11.3)

I'd like to use edge merging for data manipulation with large files (e.g. dense bipartite networks generated from secondary data where you want to cut off edge weights below some point). So I'm using 0.8 in parallel.

**Comment by eduramiba** [comment]
This issue is only about spreadsheet import. Problems in gexf improt should be included in a separate issue.

**Comment by HalldisNaamah** [comment]
I am sorry if my question seems ridiculous to you guys, but I still do not know how to get edges summed together. Has this problem been solved?

I was using the older version 0.8.2 and there it was done automatically. Now I am running 0.9.1 on Mac OS 10.10.5 and nothing happens...

I am importing CSV that contains data in this format:
Target, Source, Type
A, B, Undriected
B, C, Undirected
B, A, Unidriected
C, A, Undirected
A, B, Undirected

And after import all the edges have weight = 1.0

I have read that you need to install some updates and it should be fine but when I go to Tools -> Plugins - Check for updates it does not show any updates. And I am connected to Gephi updates Center and Gephi third Parties Plugins. 

I would appreciate your help.

**Comment by eduramiba** [comment]
Not solved yet, sorry. But it's a high priority bug.

**Comment by jeremyreffin** [comment]
Bump. This is a high priority bug for this user too! 

Thanks for all the great work.

**Comment by WouterSpekkink** [comment]
Yep, sorry to build up the pressure, but I would like to see this implemented too.
